torch/onnx/symbolic.py

```diff
@@ -574,69 +577,204 @@ def softplus(g, self, beta, threshold):
-        if ceil_mode:
-            return _unimplemented("avg_pool2d", "ceil_mode")
+        if ceil_mode and input.type().kind() != "CompleteTensorType":
+            return _unimplemented(name, "input size not accesible")
         if not stride:
             stride = kernel_size
 
         padding = tuple(tuple_fn(padding))
+        # padding to add in front in ceil_mode
+        if ceil_mode:
+            d = input.type().sizes()[-len(padding):]
+            ceil_out_dim = [math.ceil((d[i] + 2 * padding[i] - kernel_size[i]) / stride[i]) + 1
+                            for i in range(0, len(padding))]
+            padding_ceil = [(ceil_out_dim[i] - 1) * stride[i] - d[i] - 2 * padding[i] + kernel_size[i]
+                            for i in range(0, len(padding))]
+            padding_ceil_beg = (0,) * len(padding) if count_include_pad else padding
+            padding_ceil = padding_ceil_beg + tuple(padding_ceil) + tuple(tuple_fn(padding)) * 2
+
         if count_include_pad:
             input = g.op("Pad", input,
                          pads_i=((0,) * 2 + padding) * 2,
                          mode_s='constant',
                          value_f=0.)
             padding = (0,) * len(padding)
-
-        return g.op("AveragePool", input,
-                    kernel_shape_i=tuple_fn(kernel_size),
-                    strides_i=tuple_fn(stride),
-                    pads_i=padding * 2)
+        padding = ((0,) * len(padding) + padding) * 2
+        # ONNX does not support padding >=  kernel_size. Keep support padding < kernel_size,
+        # and add NaN padding to the output when padding >= kernel_size to respect the output shape
+        if ceil_mode:
+            padding = numpy.add(padding, padding_ceil)
+            padding_overflow = [1
+                                if i - len(kernel_size) in range(0, len(kernel_size))
+                                and padding[i] >= kernel_size[i - len(kernel_size)]
+                                else 0
+                                for i in range(int(len(padding) / 4), int(len(padding) / 4) + len(kernel_size))]
+            padding_overflow = (0,) * int(len(padding) / 4) * 3 + tuple(padding_overflow)
+            padding = [kernel_size[i - len(kernel_size)] - 1
+                       if i - len(kernel_size) in range(0, len(kernel_size))
+                       and padding[i] >= kernel_size[i - len(kernel_size)]
+                       else padding[i]
+                       for i in range(0, len(padding))]
+
+        output = g.op("AveragePool", input,
+                      kernel_shape_i=tuple_fn(kernel_size),
+                      strides_i=tuple_fn(stride),
+                      pads_i=padding)
+
+        if ceil_mode and padding_overflow != ((0,) * len(padding)):
+            output = g.op("Pad", output,
+                          pads_i=padding_overflow,
+                          mode_s='constant',
+                          value_f=numpy.NaN)
```

Comment:
is there a particular reason padding is NaN in pytorch instead of zero for this case?
( example : for Maxpool1D(kernel_size=3, stride=6), with the input=torch.randn(1,1,6,6), the output will have NaN values)